fendermint/vm/interpreter/src/chain.rs

```diff
@@ -173,32 +211,55 @@ where
                     }
 
                     // We can use the same result type for now, it's isomorphic.
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDown => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownExec(p) => {
+                    if !provider.is_enabled() {
+                        return Err(anyhow!("toggle on ipc top down to process messages"));
+                    }
+
```

Comment:
Something I haven't considered before but we must handle is that currently a validator can propose multiple finalities in the same block, and do so in arbitrary order. We must decide whether to allow this, and if so how to proceed during execution. 

Say the last committed finality is 100, and the next validator proposes multiple ones: [120, 110] in that order. Say we look at 120, it's final, then 110, obviously that is final too, so we vote "yes". Then during execution we do 120 first, okay, and then what about 110? We should skip it. We can check in the cache, because the previous run should have called `set_new_finality`.

The other option is to not vote if there are multiple records in the same block, which prevents ordering problems.

Reply:
Yet another option, which may be actually the easiest and cleanest to implement, is to reject backwards going finality in the contract, and then to [inspect the return value](https://github.com/consensus-shipyard/fendermint/pull/236/files#diff-38169d6d71afc04b5bc631226485edb63af1309907ec52553d956d0c55e86259R233) and if it's in an error then do not execute the cross message, and also do not update the cache. 

This avoids any potential double execution of cross-messages.

Reply:
The backward rejection is implement in the Membership PR with configuration number, see https://github.com/consensus-shipyard/ipc-solidity-actors/blob/3332383c11dfee0dc60d29a913c733d8e25752e8/src/lib/LibGateway.sol#L79.

Reply:
Okay that ignores the membership but we also need to actually reject the block hash and height, rather than go backwards.

Reply:
this one; https://github.com/consensus-shipyard/ipc-solidity-actors/blob/876b5fa56ce5dc5222589f761b8afddef41ca4f4/src/lib/LibGateway.sol#L60